src/realm/integer_flex_compressor.hpp

```diff
@@ -189,116 +190,99 @@ inline bool FlexCompressor::find_all(const Array& arr, int64_t value, size_t sta
+            v++; // We use GreaterEqual below, so this will effectively be Greater
+        }
 
-    if constexpr (!std::is_same_v<VectorCond2, NotEqual>) {
-        if (start == v_size)
-            return true;
+        if (v_width <= 16 && v_size >= 20) {
+            auto search_vector = populate(v_width, v);
+            v_start = parallel_subword_find(find_all_fields<GreaterEqual>, data, 0, v_width, compressor.msb(),
+                                            search_vector, 0, v_size);
+        }
+        else {
+            BfIterator data_iterator{data, 0, v_width, v_width, 0};
+            size_t idx = 0;
+            while (idx < v_size) {
+                if (sign_extend_field_by_mask(mask, *data_iterator) >= v) {
+                    break;
+                }
+                data_iterator.move(++idx);
+            }
+            v_start = idx;
+            /*
+            // Binary search alternative
+            v_start = lower_bound(v_size, v, mask, data_iterator);
+            */
+        }
     }
 
-    MSBs = compressor.ndx_msb();
-    search_vector = populate(ndx_width, v_start);
-    while (start < end) {
-        start = parallel_subword_find(find_all_fields_unsigned<VectorCond2>, data, offset, ndx_width, MSBs,
-                                      search_vector, start, end);
-
-        if (start < end && !state->match(start + baseindex))
-            return false;
+    if constexpr (realm::is_any_v<Cond, Equal, NotEqual>) {
+        // Check for equality.
+        if (v_start < v_size) {
+            BfIterator it{data, 0, v_width, v_width, v_start};
+            if (sign_extend_field_by_mask(mask, *it) > value) {
+                v_start = v_size; // Mark as not found
+            }
+        }
+    }
 
-        ++start;
+    if (v_start == v_size) {
+        if constexpr (realm::is_any_v<Cond, Equal, Greater>) {
+            return true; // No Matches
+        }
+        if constexpr (realm::is_any_v<Cond, NotEqual, Less>) {
+            return find_all_match(start, end, baseindex, state); // All matches
+        }
+    }
+    else if (v_start == 0) {
+        if constexpr (std::is_same_v<Cond, Less>) {
+            // No index is less than 0
+            return true; // No Matches
+        }
+        if constexpr (std::is_same_v<Cond, Greater>) {
+            // All index is greater than or equal to 0
+            return find_all_match(start, end, baseindex, state);
+        }
     }
-    return true;
-}
 
-template <typename Cond>
-inline bool FlexCompressor::run_parallel_subscan(size_t v_width, size_t v_range, size_t ndx_range)
-{
-    if constexpr (std::is_same_v<Cond, Equal> || std::is_same_v<Cond, NotEqual>) {
-        return v_width < 32 && v_range >= 20 && ndx_range >= 20;
+    // Search the indexes
+    using U = typename IndexCond<Cond>::type;
+    const auto ndx_range = end - start;
+    const auto ndx_width = compressor.ndx_width();
+    const auto v_offset = v_size * v_width;
+    if (ndx_range >= 20) {
```

Comment:
Maybe let's define a constexpr CONSTANT = 20 for this magic numbers.. Before they were contained in a single method, now it is not clear why we have this.

Reply:
👍